go/plugins/ollama/ollama.go

```diff
@@ -0,0 +1,228 @@
+raw: if true no formatting will be applied to the prompt. You may choose to use the raw parameter if you are specifying a full templated prompt in your request to the API
+keep_alive: controls how long the model will stay loaded into memory following the request (default: 5m)
+*/
+type ollamaRequest struct {
+	Messages []map[string]string `json:"messages"`
+	Model    string              `json:"model"`
+	Stream   bool                `json:"stream"`
+}
+
+// Generate makes a request to the Ollama API and processes the response.
+func (g *generator) generate(ctx context.Context, input *ai.GenerateRequest, cb func(context.Context, *ai.GenerateResponseChunk) error) (*ai.GenerateResponse, error) {
+	// Step 1: Combine parts from all messages into a single payload slice
+	var messages []map[string]string
+
+	// Add all messages to history field.
+	for _, m := range input.Messages {
+		message, err := convertParts(m.Role, m.Content)
+		if err != nil {
+			return nil, fmt.Errorf("error converting message parts: %v", err)
+		}
+		messages = append(messages, message)
+	}
+	fmt.Println("should stream", cb != nil)
+	stream := cb != nil
+	payload := ollamaRequest{
+		Messages: messages,
+		Model:    g.model,
+		Stream:   stream,
+	}
+	client := &http.Client{
+		Timeout: time.Second * 30,
+	}
+	payloadBytes, err := json.Marshal(payload)
+	req, err := http.NewRequest("POST", g.serverAddress+"/api/chat", bytes.NewBuffer(payloadBytes))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send request: %v", err)
+	}
+	defer resp.Body.Close()
+	if cb == nil {
+		// Existing behavior for non-streaming responses
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read response body: %v", err)
+		}
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("server returned non-200 status: %d, body: %s", resp.StatusCode, body)
+		}
+		response, err := translateResponse(body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse response: %v", err)
+		}
+		return response, nil
+	} else {
+		// Handle streaming response here
+		scanner := bufio.NewScanner(resp.Body) // Create a scanner to read lines
+		for scanner.Scan() {
+			line := scanner.Text()
+
+			chunk, err := translateChunk(line)
+			if err != nil {
+				// Handle parsing error (log, maybe send an error candidate?)
+				return nil, fmt.Errorf("error translating chunk: %v", err)
+			}
+			cb(ctx, chunk)
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, fmt.Errorf("error reading stream: %v", err)
+		}
+		// Handle end of stream (optional: send a final candidate to signal completion)
+	}
+	//Return an empty generate response, since we use callback for streaming
```

Comment:
In the other plugins, we wound up having the final return value be the merging of all the streaming responses we already returned (as if it were a non-streaming request). We should probably do the same here.